File: lect-player-web/src/main/frontend/src/render/rectangle.renderer.ts
import { ShapeRenderer } from "./shape.renderer";
import { Rectangle } from "../geometry/rectangle";
import { RectangleShape } from "../model/shape/rectangle.shape";

class RectangleRenderer implements ShapeRenderer {

	render(context: CanvasRenderingContext2D, shape: RectangleShape, dirtyRegion: Rectangle): void {
		const bounds = shape.bounds;

		if (bounds.isEmpty()) {
			return;
		}

		const keyEvent = shape.getKeyEvent();
		const brush = shape.brush;
		const color = brush.color.toRgba();
		const fill = keyEvent && keyEvent.altKey;

		let x = bounds.x;
		let y = bounds.y;
		let w = bounds.width;
		let h = bounds.height;

		context.beginPath();

		if (fill) {
			context.fillStyle = color;
			context.rect(x, y, w, h);
			context.fill();
		}
		else {
			context.strokeStyle = color;
			context.lineWidth = brush.width;
			context.rect(x, y, w, h);
			context.stroke();
		}

		if (shape.isSelected()) {
			if (fill) {
				// Draw the selection indicator inside the rectangle.
				x += brush.width / 2;
				y += brush.width / 2;
				w -= brush.width;
				h -= brush.width;
			}

			context.beginPath();
			context.strokeStyle = "rgb(255, 0, 100)";
			context.lineWidth = brush.width;
			context.setLineDash([5 / context.getTransformExt().getScaleX()]);
			context.rect(x, y, w, h);
			context.stroke();
		}
	}
}

export { RectangleRenderer };

File: lect-javafx/src/main/java/org/lecturestudio/javafx/render/RectangleRenderer.java
/*
 * Copyright (C) 2020 TU Darmstadt, Department of Computer Science,
 * Embedded Systems and Applications Group.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.lecturestudio.javafx.render;

import javafx.scene.canvas.GraphicsContext;
import javafx.scene.paint.Color;
import javafx.scene.shape.StrokeLineCap;
import javafx.scene.shape.StrokeLineJoin;

import org.lecturestudio.core.input.KeyEvent;
import org.lecturestudio.core.geometry.Rectangle2D;
import org.lecturestudio.core.model.shape.RectangleShape;
import org.lecturestudio.core.model.shape.Shape;
import org.lecturestudio.core.render.Renderer;
import org.lecturestudio.core.tool.Stroke;
import org.lecturestudio.javafx.beans.converter.ColorConverter;

public class RectangleRenderer implements Renderer<GraphicsContext> {

	private static final Color FOCUS_COLOR = Color.rgb(255, 255, 255, 200 / 255.0);


	@Override
	public Class<? extends Shape> forClass() {
		return RectangleShape.class;
	}

	@Override
	public void render(Shape shape, GraphicsContext context) {
		RectangleShape rectShape = (RectangleShape) shape;
		KeyEvent keyEvent = shape.getKeyEvent();
		Stroke stroke = rectShape.getStroke();
		double width = stroke.getWidth();

		Rectangle2D bounds = shape.getBounds().clone();
		Color color = ColorConverter.INSTANCE.to(stroke.getColor());

		if (keyEvent != null && keyEvent.isAltDown()) {
			context.setFill(color);
			context.fillRect(bounds.getX(), bounds.getY(), bounds.getWidth(), bounds.getHeight());
		}
		else {
			context.setLineWidth(width);
			context.setStroke(color);
			context.strokeRect(bounds.getX(), bounds.getY(), bounds.getWidth(), bounds.getHeight());
		}

		// Focus
		if (shape.isSelected()) {
			if (keyEvent != null && keyEvent.isAltDown()) {
				context.setFill(FOCUS_COLOR);
				context.fillRect(bounds.getX() + width, bounds.getY() + width, bounds.getWidth() - 2 * width, bounds.getHeight() - 2 * width);
			}
			else {
				double penWidth = width * 0.5;

				context.setLineCap(StrokeLineCap.ROUND);
				context.setLineJoin(StrokeLineJoin.ROUND);
				context.setLineWidth(penWidth);
				context.setStroke(FOCUS_COLOR);
				context.strokeRect(bounds.getX(), bounds.getY(), bounds.getWidth(), bounds.getHeight());
			}
		}
	}
}
